pmkoalas/models/petrinet.py

```diff
@@ -0,0 +1,341 @@
+
+# 
+# Weighted Petri net with place and transition labels.
+# 
+# Allows explicit place or transition ids for simpler comparison, especially
+# during testing.
+#
+
+import operator
+import xml.etree.ElementTree as ET
+
+ENCODING='unicode'
+
+def steq(self,other):
+    if type(other) is type(self):
+        return self.__dict__ == other.__dict__
+    return False
+
+
+
+class LabelledPetriNet:
+    def __init__(self,places,transitions,arcs,label=None):
+        self._places = places
+        self._transitions = transitions
+        self._arcs = arcs
+        self._label = label
+
+    @property 
+    def places(self):
+        return self._places
+
+    @property
+    def transitions(self):
+        return self._transitions
+
+    @property
+    def arcs(self):
+        return self._arcs
+
+    @property
+    def label(self):
+        return self._label
+
+    def __eq__(self,other):
+        if isinstance(other,self.__class__):
+            return self._label  == other._label and \
+                   self._places == other._places and \
+                   self._transitions == other._transitions and \
+                   self._arcs   == self._arcs
+        return False
+
+    def reprcontents(self):
+        return f"places: {self._places} transitions: {self._transitions} arcs: {self._arcs}"
+
+    def __repr__(self):
+        return f"LabelledPetriNet:{self._label} " + self.reprcontents()
+
+    def __str__(self):
+        return "LabelledPetriNet(" + self._label + ")\n" \
+            +  " Places: " + str(self.places) + "\n" \
+            +  " Trans: " + str(self.transitions) + "\n" \
+            +  " Arcs: " + str(self.arcs)
+    
+
+class MutableLabelledPetriNet(LabelledPetriNet):
```

Comment:
This might seem a bit odd, but there is nothing private in python. Do we really need this class?

Reply:
The functional extras that this class provides already exists, I believe.

Reply:
```python
from pmkoalas.dtlog import convert
from pmkoalas.discovery.alpha_miner import AlphaMinerInstance
from pmkoalas.models.petrinet import LabelledPetriNet

# create a log
log = convert( 
    "a b e d z",
    "a c e f z",
    "a y e g z"
)

# run alpha miner 
miner = AlphaMinerInstance()
places, transitions, flows = miner.mine_model(log)

# use given places, transitions and flows to create an 
# instance of a petri net 
net = LabelledPetriNet(places, transitions, flows, "hello")
# display net in some potential human readable manner
net._places.add("p3")
print(net)
```

Reply:
In fact, I can do this in many ways, see below (very interesting) : 
```python

from pmkoalas.dtlog import convert
from pmkoalas.discovery.alpha_miner import AlphaMinerInstance
from pmkoalas.models.petrinet import LabelledPetriNet

# create a log
log = convert( 
    "a b e d z",
    "a c e f z",
    "a y e g z"
)

# run alpha miner 
miner = AlphaMinerInstance()
places, transitions, flows = miner.mine_model(log)

# use given places, transitions and flows to create an 
# instance of a petri net 
net = LabelledPetriNet(places, transitions, flows, "hello")
# display net in some potential human readable manner
net._places.add("p3")
print(net)
print()

# or i could do this
places.add("p4")
print("printing initial set")
print(places)
print()
print("printing changed net")
print(net)
print()

# or this 
net.places.add("p5")
print(net)
```

Reply:
Immutability (even if enforced by convention here) is useful in the implementation of the discovery algo. I didn't include it as I don't think obscure algos should be in the package. Specifically it uses frozensets at some points. I don't think mutability is a desirable property in a default PN impl.

Reply:
You are reusing the references, adam; you have not enforced immutability by convention. Any collection in python is passed by reference, including sets. If you wish to enforce immutability then you must create a new collection in __init__ and all properties must return copies of the collections.

Reply:
However, your comment below seems to indicate that the structure is pointless, other than to be a wrapper container. So why bother with any of these considerations? Just make it a pointless wrapper.